crates/wit-parser/src/resolve.rs

```diff
@@ -1873,17 +1929,48 @@ impl Remap {
         if let Some(spans) = spans {
             assert_eq!(iface.functions.len(), spans.funcs.len());
         }
-        for (i, (_, func)) in iface.functions.iter_mut().enumerate() {
-            if !resolve.include_stability(&func.stability) {
+        for (i, (func_name, func)) in iface.functions.iter_mut().enumerate() {
+            if !resolve
+                .include_stability(&func.stability, iface_pkg_id)
+                .with_context(|| {
+                    format!(
+                        "failed to process feature gate for function [{func_name}] in package [{}]",
+                        resolve.packages[*iface_pkg_id].name,
+                    )
+                })?
+            {
                 continue;
             }
             let span = spans.map(|s| s.funcs[i]);
             self.update_function(resolve, func, span)
                 .with_context(|| format!("failed to update function `{}`", func.name))?;
         }
-        iface
+
+        // Filter out all of the existing functions in interface which fail the
+        // `include_stability()` check, as they shouldn't be available.
+        iface.functions = iface
             .functions
-            .retain(|_, f| resolve.include_stability(&f.stability));
+            .drain(..)
+            .map(|(name, f)| {
+                resolve
+                    .include_stability(&f.stability, iface_pkg_id)
+                    .map(|include| (name, f, include))
+            })
+            .collect::<Result<Vec<(String, Function, _)>>>()
+            .with_context(|| {
+                format!(
+                    "error while checking feature gates of functions in iface [{}]",
+                    iface
+                        .name
+                        .as_ref()
+                        .map(String::as_str)
+                        .unwrap_or("<unknown>"),
+                )
+            })?
+            .into_iter()
+            .filter_map(|(name, func, include)| include.then_some((name, func)))
+            .collect::<IndexMap<String, Function>>();
```

Comment:
This whole combinator might be a bit easier to read as an imperative loop perhaps?

```rust
for (name, func) in mem::take(&mut iface.functions) {
    // include_stability + with_context
    iface.functions.insert(name, func);
}
```

Reply:
OK will use that -- agree that it would be a bit easier to read that way

- [ ] rewrite iface fn update code